src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -151,3 +152,38 @@ def signature_hash_algorithm(self):
             raise UnsupportedAlgorithm(
                 "Signature algorithm OID:{0} not recognized".format(oid)
             )
+
+    @property
+    def extensions(self):
+        extensions = []
+        seen_oids = set()
+        extcount = self._backend._lib.X509_get_ext_count(self._x509)
+        for i in range(0, extcount):
+            ext = self._backend._lib.X509_get_ext(self._x509, i)
+            assert ext != self._backend._ffi.NULL
+            crit = self._backend._lib.X509_EXTENSION_get_critical(ext)
+            critical = crit == 1
+            oid = x509.ObjectIdentifier(_obj2txt(self._backend, ext.object))
+            if oid in seen_oids:
+                raise x509.DuplicateExtension(
+                    "Duplicate {0} extension found".format(oid)
+                )
+            elif oid == x509.OID_BASIC_CONSTRAINTS and critical:
+                # TODO: remove this obviously.
+                warnings.warn(
+                    "Extension support is not fully implemented. A basic "
+                    "constraints extension with the critical flag was seen and"
+                    " IGNORED."
+                )
+                # This will move to the else block in the next PR
+                seen_oids.add(oid)
```

Comment:
k

On Sat, Mar 28, 2015 at 11:55 AM, Paul Kehrer <notifications@github.com>
wrote:

> In src/cryptography/hazmat/backends/openssl/x509.py
> <https://github.com/pyca/cryptography/pull/1793#discussion_r27345980>:
>
> > +            crit = self._backend._lib.X509_EXTENSION_get_critical(ext)
> > +            critical = crit == 1
> > +            oid = x509.ObjectIdentifier(_obj2txt(self._backend, ext.object))
> > +            if oid in seen_oids:
> > +                raise x509.DuplicateExtension(
> > +                    "Duplicate {0} extension found".format(oid)
> > +                )
> > +            elif oid == x509.OID_BASIC_CONSTRAINTS and critical:
> > +                # TODO: remove this obviously.
> > +                warnings.warn(
> > +                    "Extension support is not fully implemented. A basic "
> > +                    "constraints extension with the critical flag was seen and"
> > +                    " IGNORED."
> > +                )
> > +                # This will move to the else block in the next PR
> > +                seen_oids.add(oid)
>
> That's fine for this PR, but I'll have to add continue back in next PR
> since we'll be actually building an extension in this loop at that point.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/1793/files#r27345980>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084